# Home Work 2. Pandas practice.

By the end of this task you will be able to manipulate huge tabular data:
1. Select observations/features by condition/index;
2. Compute different column's statistics (min, max, mean, quantiles etc.);
3. Create new columns (feature engineering);

and more.

Useful links with `pandas` tutorials:

1. Single notebook, covers basic pandas functionality (starting with renaming columns ending with using map, apply etc) ~ 30 short examples with links on videos https://nbviewer.jupyter.org/github/justmarkham/pandas-videos/blob/master/pandas.ipynb . Highly recommended for everyone. (about 1-3 hours to go through)

2. https://github.com/guipsamora/pandas_exercises/ 11 topics covering all essential functionality.

This task will be an easy ride after these tutorials.

We are using house sale price data from King County, Wahington, USA. This dataset is in public domain and can be obtained from Kaggle: https://www.kaggle.com/harlfoxem/housesalesprediction

You need to place `kc_house_data.csv` file in the same directory as this notebook.

In [1]:
import pandas as pd

# 0. Loading data

As always in Data Science you are starting with making nice cup of tea (or coffee). Your next move is to load the data:

Methods to use:
* `pd.read_csv()`

In [2]:
# Load the data
house_data = pd.read_csv("kc_house_data.csv", index_col='id')

# 1. Data inspection
Methods to use:
* `df.head()`
* `df.tail()`
* `df.columns`
* `df.shape`

## Q1.0 
Display first 10 observations.

In [3]:
#YOUR CODE HERE

date      price  bedrooms  bathrooms  sqft_living   
id                                                                         
7129300520  20141013T000000   221900.0         3       1.00         1180  \
6414100192  20141209T000000   538000.0         3       2.25         2570   
5631500400  20150225T000000   180000.0         2       1.00          770   
2487200875  20141209T000000   604000.0         4       3.00         1960   
1954400510  20150218T000000   510000.0         3       2.00         1680   
7237550310  20140512T000000  1225000.0         4       4.50         5420   
1321400060  20140627T000000   257500.0         3       2.25         1715   
2008000270  20150115T000000   291850.0         3       1.50         1060   
2414600126  20150415T000000   229500.0         3       1.00         1780   
3793500160  20150312T000000   323000.0         3       2.50         1890   

            sqft_lot  floors  waterfront  view  condition  grade  sqft_above   
id                                                                             
7129300520      5650     1.0           0     0          3      7        1180  \
6414100192      7242     2.0           0     0          3      7        2170   
5631500400     10000     1.0           0     0          3      6         770   
2487200875      5000     1.0           0     0          5      7        1050   
1954400510      8080     1.0           0     0          3      8        1680   
7237550310    101930     1.0           0     0          3     11        3890   
1321400060      6819     2.0           0     0          3      7        1715   
2008000270      9711     1.0           0     0          3      7        1060   
2414600126      7470     1.0           0     0          3      7        1050   
3793500160      6560     2.0           0     0          3      7        1890   

            sqft_basement  yr_built  yr_renovated  zipcode      lat     long   
id                                                                             
7129300520              0      1955             0    98178  47.5112 -122.257  \
6414100192            400      1951          1991    98125  47.7210 -122.319   
5631500400              0      1933             0    98028  47.7379 -122.233   
2487200875            910      1965             0    98136  47.5208 -122.393   
1954400510              0      1987             0    98074  47.6168 -122.045   
7237550310           1530      2001             0    98053  47.6561 -122.005   
1321400060              0      1995             0    98003  47.3097 -122.327   
2008000270              0      1963             0    98198  47.4095 -122.315   
2414600126            730      1960             0    98146  47.5123 -122.337   
3793500160              0      2003             0    98038  47.3684 -122.031   

            sqft_living15  sqft_lot15  
id                                     
7129300520           1340        5650  
6414100192           1690        7639  
5631500400           2720        8062  
2487200875           1360        5000  
1954400510           1800        7503  
7237550310           4760      101930  
1321400060           2238        6819  
2008000270           1650        9711  
2414600126           1780        8113  
3793500160           2390        7570

## Q1.1
What is the price of a house with `id` == 7237550310?

In [4]:
#YOUR CODE HERE

1225000.0


## Q1.2
When was the house with `id` == 2414600126 built (`yr_built`)?

In [5]:
#YOUR CODE HERE

1960


## Q1.3
Observe last 10 rows.

In [6]:
#YOUR CODE HERE

date      price  bedrooms  bathrooms  sqft_living   
id                                                                         
7852140040  20140825T000000   507250.0         3       2.50         2270  \
9834201367  20150126T000000   429000.0         3       2.00         1490   
3448900210  20141014T000000   610685.0         4       2.50         2520   
7936000429  20150326T000000  1007500.0         4       3.50         3510   
2997800021  20150219T000000   475000.0         3       2.50         1310   
263000018   20140521T000000   360000.0         3       2.50         1530   
6600060120  20150223T000000   400000.0         4       2.50         2310   
1523300141  20140623T000000   402101.0         2       0.75         1020   
291310100   20150116T000000   400000.0         3       2.50         1600   
1523300157  20141015T000000   325000.0         2       0.75         1020   

            sqft_lot  floors  waterfront  view  condition  grade  sqft_above   
id                                                                             
7852140040      5536     2.0           0     0          3      8        2270  \
9834201367      1126     3.0           0     0          3      8        1490   
3448900210      6023     2.0           0     0          3      9        2520   
7936000429      7200     2.0           0     0          3      9        2600   
2997800021      1294     2.0           0     0          3      8        1180   
263000018       1131     3.0           0     0          3      8        1530   
6600060120      5813     2.0           0     0          3      8        2310   
1523300141      1350     2.0           0     0          3      7        1020   
291310100       2388     2.0           0     0          3      8        1600   
1523300157      1076     2.0           0     0          3      7        1020   

            sqft_basement  yr_built  yr_renovated  zipcode      lat     long   
id                                                                             
7852140040              0      2003             0    98065  47.5389 -121.881  \
9834201367              0      2014             0    98144  47.5699 -122.288   
3448900210              0      2014             0    98056  47.5137 -122.167   
7936000429            910      2009             0    98136  47.5537 -122.398   
2997800021            130      2008             0    98116  47.5773 -122.409   
263000018               0      2009             0    98103  47.6993 -122.346   
6600060120              0      2014             0    98146  47.5107 -122.362   
1523300141              0      2009             0    98144  47.5944 -122.299   
291310100               0      2004             0    98027  47.5345 -122.069   
1523300157              0      2008             0    98144  47.5941 -122.299   

            sqft_living15  sqft_lot15  
id                                     
7852140040           2270        5731  
9834201367           1400        1230  
3448900210           2520        6023  
7936000429           2050        6200  
2997800021           1330        1265  
263000018            1530        1509  
6600060120           1830        7200  
1523300141           1020        2007  
291310100            1410        1287  
1523300157           1020        1357

## Q1.4
How many floors house with `id` == 2997800021 has?

In [7]:
#YOUR CODE HERE

2.0


## Q1.5
What is the price of a house with `id` == 263000018?

In [8]:
#YOUR CODE HERE

360000.0


## Q1.6
How many observations (rows) are in the data?

In [9]:
#YOUR CODE HERE

21613


## Q1.7
How many columns are in the data?

In [10]:
#YOUR CODE HERE

20


##  Q1.8 
Print all the columns names

In [11]:
#YOUR CODE HERE

Index(['date', 'price', 'bedrooms', 'bathrooms', 'sqft_living', 'sqft_lot',
       'floors', 'waterfront', 'view', 'condition', 'grade', 'sqft_above',
       'sqft_basement', 'yr_built', 'yr_renovated', 'zipcode', 'lat', 'long',
       'sqft_living15', 'sqft_lot15'],
      dtype='object')


# 2. Basic data exploration

Methods to use:
* `.count()` number of not NaN's in every column.
* `.nunique()` Count number of unique values in every column . 
* `.unique()`  return unique values unique values in a column as list. 
* `.min()`, `.max()`, `.mean()`, `.std()` to display specific statistics about the data.
* `.isnull()` return True if value is NaN (missing) and False if its not.


## Q2.0 
Count number of unique values in every column.

In [12]:
#YOUR CODE HERE

date              372
price            4028
bedrooms           13
bathrooms          30
sqft_living      1038
sqft_lot         9782
floors              6
waterfront          2
view                5
condition           5
grade              12
sqft_above        946
sqft_basement     306
yr_built          116
yr_renovated       70
zipcode            70
lat              5034
long              752
sqft_living15     777
sqft_lot15       8689
dtype: int64

## Q2.1 
Display all the unique values in column `condition`.

In [13]:
#YOUR CODE HERE

array([3, 5, 4, 1, 2])

## Q2.2
How many unique values are in the `yr_renovated` column?

In [14]:
#YOUR CODE HERE

70


## Q2.3
How many unique values are in the `long` column?

In [15]:
#YOUR CODE HERE

752


## Q2.4
Display number of NaN's in every column. 

**Hint:** combine `.isnull()` and `.sum()` methods.

In [16]:
#YOUR CODE HERE

date             0
price            0
bedrooms         0
bathrooms        0
sqft_living      0
sqft_lot         0
floors           0
waterfront       0
view             0
condition        0
grade            0
sqft_above       0
sqft_basement    0
yr_built         0
yr_renovated     0
zipcode          0
lat              0
long             0
sqft_living15    0
sqft_lot15       0
dtype: int64

# 3. Data selection

In pandas.DataFrame you could select

1. Row/s by position (integer number \[0 .. number of rows - 1\]) `.iloc` or by DataFrame.index `.loc`:
```
data.loc[0]
data.loc[5:10]
data.iloc[0]
data.iloc[5:10]
```
*Though, this is probably the worst way to manipulate rows.*

2. Columns by name
```
data[columname]
```
3. Row/s and columns
```
data.loc[10, columname]
data.iloc[10, columname]
```
4. Using boolean mask
```
mask = data[columname] > value
data[mask]
```
You could combine multiple conditions using `&` or `|` (and, or)

```
cond1 = data[columname1] > value1
cond2 = data[columname2] > value2
data[cond1 & cond2]
```
5. Using queries `.query()`:
```
value = 5
data.query("columname > value")
```
You could combine multiple conditions using `and`, `or`

```
data.query("(columname1 > value1) and (columname2 > value2)")
```
and others. See https://pandas.pydata.org/pandas-docs/stable/user_guide/indexing.html for more examples.

*Remember to use different quotation marks " or ' for columnname inside a query.*

## Q3.0
How many bathrooms have a house on row 314?


In [17]:
#YOUR CODE HERE

3.25

## Q3.1
What is the grade of a house on row 2718?

In [18]:
#YOUR CODE HERE

7

## Q3.2
How many times the house with index 1000102 were sold? Display all the rows.

In [19]:
#YOUR CODE HERE

date     price  bedrooms  bathrooms  sqft_living   
id                                                                     
1000102  20140916T000000  280000.0         6        3.0         2400  \
1000102  20150422T000000  300000.0         6        3.0         2400   

         sqft_lot  floors  waterfront  view  condition  grade  sqft_above   
id                                                                          
1000102      9373     2.0           0     0          3      7        2400  \
1000102      9373     2.0           0     0          3      7        2400   

         sqft_basement  yr_built  yr_renovated  zipcode      lat     long   
id                                                                          
1000102              0      1991             0    98002  47.3262 -122.214  \
1000102              0      1991             0    98002  47.3262 -122.214   

         sqft_living15  sqft_lot15  
id                                  
1000102           2060        7316  
1000102           2060        7316

## Q3.3
When was built the house with index 104510440?

In [20]:
#YOUR CODE HERE

1984

# 4. Columns statistics
Methods to use:
* `.value_counts()` to count number of unique values in a specific column.
* `.describe()` to display basic statistic about the data.
* `.dtypes` field to display data types in columns.

## Q4.0
For every unique `view` value give its number of occurences. Hint: first you need to select column `view` and then use `.value_counts()`

In [21]:
#YOUR CODE HERE

view
0    19489
2      963
3      510
1      332
4      319
Name: count, dtype: int64

## Q4.1
For every unique `condition` value give its number of occurences.

In [22]:
#YOUR CODE HERE

condition
3    14031
4     5679
5     1701
2      172
1       30
Name: count, dtype: int64

## Q4.2
What are the max, min, mean and the std of the `bedrooms` column? 

**Hint:** try to combine `describe`, `.loc()` and single column selection.

In [23]:
#YOUR CODE HERE

max     33.000000
min      0.000000
mean     3.370842
std      0.930062
Name: bedrooms, dtype: float64

## Q4.3
What are the `25%` (25th percentile), `50%` (50th percentile) and `max` value of the `sqft_living` column?

In [24]:
#YOUR CODE HERE

25%     1427.0
50%     1910.0
max    13540.0
Name: sqft_living, dtype: float64

## Q4.4 
Display data types of all columns (int)

In [25]:
#YOUR CODE HERE

date              object
price            float64
bedrooms           int64
bathrooms        float64
sqft_living        int64
sqft_lot           int64
floors           float64
waterfront         int64
view               int64
condition          int64
grade              int64
sqft_above         int64
sqft_basement      int64
yr_built           int64
yr_renovated       int64
zipcode            int64
lat              float64
long             float64
sqft_living15      int64
sqft_lot15         int64
dtype: object

## Q4.5 
What are the columns with dtype == `float64`? Display all of them.

In [26]:
#YOUR CODE HERE

price        float64
bathrooms    float64
floors       float64
lat          float64
long         float64
dtype: object

# 5. Logical conditions

## Q5.0
When was built (`yr_build`) the only house with basement area (`sqft_basement`) = 1024 sqft?

In [27]:
#YOUR CODE HERE

id
4136980090    2006
Name: yr_built, dtype: int64

## Q5.1
How many houses were sold (`price`) for 256000 dollars? Display all of them.

In [28]:
#YOUR CODE HERE

date     price  bedrooms  bathrooms  sqft_living   
id                                                                        
7211400990  20150303T000000  256000.0         2       1.00          860  \
8100000090  20141111T000000  256000.0         3       2.50         1480   
7224000315  20140723T000000  256000.0         3       1.00          880   
9352901085  20150204T000000  256000.0         3       1.00         1290   
4058801230  20150305T000000  256000.0         4       1.75         1270   
8001100030  20150102T000000  256000.0         3       2.50         1570   
873900240   20140702T000000  256000.0         4       2.50         2050   
603001045   20150429T000000  256000.0         3       1.00          950   

            sqft_lot  floors  waterfront  view  condition  grade  sqft_above   
id                                                                             
7211400990      5000     1.0           0     0          3      6         860  \
8100000090      7200     2.0           0     0          5      7        1480   
7224000315      5375     1.0           0     0          4      5         880   
9352901085      4720     1.0           0     0          3      6         790   
4058801230      6825     1.0           0     2          3      7        1270   
8001100030      5113     1.0           0     0          3      7        1090   
873900240       5787     2.0           0     0          3      7        2050   
603001045       4000     1.0           0     0          3      7         780   

            sqft_basement  yr_built  yr_renovated  zipcode      lat     long   
id                                                                             
7211400990              0      1915          1945    98146  47.5133 -122.356  \
8100000090              0      1995             0    98010  47.3126 -122.023   
7224000315              0      1924             0    98055  47.4858 -122.202   
9352901085            500      1948             0    98106  47.5186 -122.358   
4058801230              0      1950             0    98178  47.5051 -122.242   
8001100030            480      1996             0    98001  47.3327 -122.290   
873900240               0      2002             0    98198  47.3527 -122.315   
603001045             170      1949             0    98118  47.5232 -122.284   

            sqft_living15  sqft_lot15  
id                                     
7211400990           1220        5000  
8100000090           1350        7200  
7224000315            980        4838  
9352901085           1110        4720  
4058801230           1800        6930  
8001100030           1570        5150  
873900240            2030        6615  
603001045            1230        4000

## Q5.2
What was the price of a house with 5 `bathrooms`, built (`yr_built`) in 1998 and `grade` with 10 score?

In [29]:
#YOUR CODE HERE

id
2322069010    1180000.0
Name: price, dtype: float64

## Q5.3
Display all the houses with 6 `bedrooms` and less than 2000 `sqft_living` area.

In [30]:
#YOUR CODE HERE

date     price  bedrooms  bathrooms  sqft_living   
id                                                                        
6190701146  20150415T000000  520500.0         6       2.50         1880  \
2724069117  20141114T000000  380000.0         6       2.00         1870   
9276202190  20140808T000000  545000.0         6       1.75         1820   
1823049202  20140610T000000  175000.0         6       1.50         1930   
1823049202  20150107T000000  326000.0         6       1.50         1930   
809000945   20150106T000000  563000.0         6       1.00         1730   
7227800110  20150409T000000  315000.0         6       2.00         1750   
1443500305  20141013T000000  194990.0         6       2.50         1560   
862000020   20150206T000000  800000.0         6       1.00         1430   
3885808035  20150316T000000  619500.0         6       1.50         1680   
2568300045  20140625T000000  305000.0         6       2.00         1900   
2568300045  20150319T000000  649950.0         6       2.00         1900   
2224700045  20140804T000000  375000.0         6       2.00         1900   
823059185   20140520T000000  326000.0         6       3.00         1880   
7942602080  20140616T000000  660000.0         6       1.75         1840   
5132000140  20140618T000000  175000.0         6       1.00         1370   
5132000140  20150120T000000  415000.0         6       1.00         1370   
6746701090  20140619T000000  680000.0         6       2.00         1670   
7224000450  20141230T000000  230000.0         6       1.50         1810   
2887703155  20150225T000000  642000.0         6       1.00         1530   

            sqft_lot  floors  waterfront  view  condition  grade  sqft_above   
id                                                                             
6190701146     14350     1.0           0     0          5      7        1640  \
2724069117      6969     1.0           0     0          4      7        1870   
9276202190      6250     1.0           0     0          3      7        1130   
1823049202      8400     1.0           0     0          3      7        1030   
1823049202      8400     1.0           0     0          3      7        1030   
809000945       2760     1.5           0     0          3      7        1250   
7227800110     17685     1.0           0     0          4      5        1750   
1443500305      7144     1.0           0     0          3      6        1060   
862000020      20620     2.0           0     0          3      7        1430   
3885808035      5202     1.5           0     0          2      7        1680   
2568300045      8240     1.0           0     0          2      7        1200   
2568300045      8240     1.0           0     0          2      7        1200   
2224700045      8057     1.0           0     0          4      7        1170   
823059185       7200     1.0           0     0          4      7        1880   
7942602080      2774     1.0           0     0          3      7        1060   
5132000140      5080     1.5           0     0          3      6        1120   
5132000140      5080     1.5           0     0          3      6        1120   
6746701090      3000     1.0           0     0          5      7         900   
7224000450      4838     1.5           0     0          4      5        1050   
2887703155      4305     1.5           0     0          4      7        1530   

            sqft_basement  yr_built  yr_renovated  zipcode      lat     long   
id                                                                             
6190701146            240      1955             0    98133  47.7556 -122.352  \
2724069117              0      1968             0    98027  47.5342 -122.036   
9276202190            690      1954             0    98116  47.5790 -122.390   
1823049202            900      1971             0    98146  47.4869 -122.340   
1823049202            900      1971             0    98146  47.4869 -122.340   
809000945             480      1918             0  

## Q5.4
What is the mean number of `bathrooms` in houses with `grade` above 9 (10 and more)?

In [31]:
#YOUR CODE HERE

3.197432762836186

## Q5.5
What was the `price` of the least expensive house sold, built (`yr_built`) between 1991 and 2000?

In [32]:
#YOUR CODE HERE

154000.0

# 6. Creating new columns


Creating new column of pandas.DataFrame is as easy as:
```
data['new_awesome_column'] = [] 
```
that's it. But such a column is relatively useless. 
Typically, you would compute something new based on existing data and save it in a new column. 
For example one might want to compute total area of the house as a sum of all `sqft_` columns, or
create a boolean column of whether the house has `grade` > 2 or anything else:

```
data['total_area'] = data[col1] + data[col2] + ...
data['high_value'] = data[col] > 5
```

## Q6.0
Create a sqft_tot_area column (sum of all columns with sqft_ prefix) using any method above. 
You should sum: 
* `sqft_living`
* `sqft_lot`
* `sqft_above`
* `sqft_basement`
* `sqft_living15`
* `sqft_lot15`

Display all of then  (6 old columns and 1 newly created).

In [33]:
#YOUR CODE HERE

sqft_living  sqft_lot  sqft_above  sqft_basement  sqft_living15   
id                                                                            
7129300520         1180      5650        1180              0           1340  \
6414100192         2570      7242        2170            400           1690   
5631500400          770     10000         770              0           2720   
2487200875         1960      5000        1050            910           1360   
1954400510         1680      8080        1680              0           1800   

            sqft_lot15  sqft_tot_area  
id                                     
7129300520        5650          15000  
6414100192        7639          21711  
5631500400        8062          22322  
2487200875        5000          15280  
1954400510        7503          20743

## Q6.1
Create a new column `sqm_tot_area` using `sqft_tot_area` and the fact that 1 foot = 0.3048 meters. 

Display first 5 values for `sqft_tot_area`, and `sqm_tot_area` columns.

In [34]:
#YOUR CODE HERE

sqft_tot_area  sqm_tot_area
id                                     
7129300520          15000     4572.0000
6414100192          21711     6617.5128
5631500400          22322     6803.7456
2487200875          15280     4657.3440
1954400510          20743     6322.4664

## Q6.2
Create a new column `sqm_aver_floor_area` by dividing `sqm_tot_area` by number of `floors`. 

Display first 5 values for `sqm_tot_area`, `floors`, `sqm_aver_floor_area` columns.

In [35]:
#YOUR CODE HERE

sqm_tot_area  floors  sqm_aver_floor_area
id                                                   
7129300520     4572.0000     1.0            4572.0000
6414100192     6617.5128     2.0            3308.7564
5631500400     6803.7456     1.0            6803.7456
2487200875     4657.3440     1.0            4657.3440
1954400510     6322.4664     1.0            6322.4664

## Q6.3
Create a new bool column `high_class` it is True if the house has `grade` >= 9 and `condition` >= 4. Display first 10 values for new column.

In [36]:
#YOUR CODE HERE

id
7129300520    False
6414100192    False
5631500400    False
2487200875    False
1954400510    False
7237550310    False
1321400060    False
2008000270    False
2414600126    False
3793500160    False
Name: high_class, dtype: bool

# Q7. Final steps
1. Make sure your .ipynb is linearly executable.
2. Restart & Run All.
3. No ERROR cells.
4. Save your .ipynb file: Name_Surname_HW2.ipynb
5. Push your solution to a personal GitHub repo.
6. Share link to the solution via email.